packages/concerto-tools/lib/codegen/fromcto/avro/avrovisitor.js

```diff
@@ -0,0 +1,250 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const util = require('util');
+
+/**
+ * Convert the contents of a ModelManager to Avro IDL code.
+ * All generated code is placed into the 'Protocol' protocol. Set a
+ * fileWriter property (instance of FileWriter) on the parameters
+ * object to control where the generated code is written to disk.
+ *
+ * @private
+ * @class
+ * @memberof module:concerto-tools
+ */
+class AvroVisitor {
+    /**
+     * Visitor design pattern
+     * @param {Object} thing - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @public
+     */
+    visit(thing, parameters) {
+        if (thing.isModelManager?.()) {
+            return this.visitModelManager(thing, parameters);
+        } else if (thing.isModelFile?.()) {
+            return this.visitModelFile(thing, parameters);
+        } else if (thing.isEnum?.()) {
+            return this.visitEnumDeclaration(thing, parameters);
+        } else if (thing.isClassDeclaration?.()) {
+            return this.visitClassDeclaration(thing, parameters);
+        } else if (thing.isField?.()) {
+            return this.visitField(thing, parameters);
+        } else if (thing.isRelationship?.()) {
+            return this.visitRelationship(thing, parameters);
+        } else if (thing.isEnumValue?.()) {
+            return this.visitEnumValueDeclaration(thing, parameters);
+        } else {
+            throw new Error(
+                'Unrecognised type: ' +
+                    typeof thing +
+                    ', value: ' +
+                    util.inspect(thing, {
+                        showHidden: true,
+                        depth: 2,
+                    })
+            );
+        }
+    }
+
+    /**
+     * Visitor design pattern
+     * @param {ModelManager} modelManager - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitModelManager(modelManager, parameters) {
+        modelManager.getModelFiles(true).forEach((modelFile) => {
+            modelFile.accept(this, parameters);
+        });
+        return null;
+    }
+
+    /**
+     * Visitor design pattern
+     * @param {ModelFile} modelFile - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitModelFile(modelFile, parameters) {
+        parameters.fileWriter.openFile(modelFile.getNamespace() + '.avdl');
+
+        parameters.fileWriter.writeLine(
+            0,
+            `@namespace("${modelFile.getNamespace()}")`
+        );
+        parameters.fileWriter.writeLine(0, 'protocol MyProtocol {\n');
```

Comment:
Could we make this an option that we pass in via the CLI?

Reply:
Added a parameter.